### PR TITLE
[Integration] Modify dirac-admin-get-pilot-output to get remote pilot log 

### PR DESCRIPTION
**Pull request by martynia**
BEGINRELEASENOTES

*WMS
NEW: Enhance dirac-admin-get-pilot-output to download remote pilot logs. Currently only from a SE.

ENDRELEASENOTES

Changes are:

1. `dirac_admin_get_pilot_output.py` - add a -r switch to get remote pilot logs rather than classic ones.
2. `DiracAdmin.py` to call a service method for remote logs
3. `PilotManagerHandler.py` to load a plugin for pilot logs retrieval.
4. `FileCacheDownloadPlugin.py` a plugin to download a log file and return it to a client

### CODE REVIEW

File: src/DIRAC/Interfaces/API/DiracAdmin.py [hunk 1]
```diff
@@ -483,14 +485,17 @@ def getJobPilotOutput(self, jobID, directory=""):
         return result
 
     #############################################################################
-    def getPilotOutput(self, gridReference, directory=""):
+    def getPilotOutput(self, gridReference, directory="", remote=False):
         """Retrieve the pilot output  (std.out and std.err) for an existing job in the WMS.
```
Comment (fstagni) [inline src/DIRAC/Interfaces/API/DiracAdmin.py]:
```suggestion
        """Retrieve the pilot output  (std.out and std.err) for an existing pilot reference.
```

File: src/DIRAC/Interfaces/API/DiracAdmin.py [hunk 2]
```diff
@@ -431,7 +431,7 @@ def resetJob(self, jobID):
         return result
 
     #############################################################################
-    def getJobPilotOutput(self, jobID, directory=""):
+    def getJobPilotOutput(self, jobID, directory="", remote=False):
```
Comment (fstagni) [inline src/DIRAC/Interfaces/API/DiracAdmin.py]: Somehow it does not feel right that user would have to decide about this (seems too low level to be exposed at API level)

Comment (martynia) [reply src/DIRAC/Interfaces/API/DiracAdmin.py]: What would be an other option? If it is used in an Dirac-admin command then it is an expert route. The intention was to have a classic log retrieval mechanism a default one, and the external log retriever only used in special cases, no?

**Comment by fstagni** [comment]
You would need to rebase and fix the conflicts.

**Comment by martynia** [comment]
I rebased, but there are some strange errors I don't recognise...

File: src/DIRAC/Interfaces/API/DiracAdmin.py (hunk 2 above)
Comment (fstagni) [reply src/DIRAC/Interfaces/API/DiracAdmin.py]: Then things can be tried in this order?

Comment (martynia) [reply src/DIRAC/Interfaces/API/DiracAdmin.py]: We could try to get logs in a specific order: classic then remote. The specific order could even be configurable on the server. The point is, that if someone ever wants to test the remote log retrieval and the classic logs come first, then there is no way to get the remote logs by using a `dirac-admin` command line if we drop the `-r` flag.

So my proposal is:

- remove the `-r` flag
- set the default order to classic/remote
- add a config flag to the CS `(remoteLogsPriority=True)` or similar. Default `False.`

File: src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    """
+    Class to handle log file download from an SE
+    """
+
+    def __init__(self):
+        """
+        Sets the pilot log files location for a WebServer.
+
+        """
+        self.setup = gConfig.getValue("/DIRAC/Setup", None)
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py]: Remove this, I don't think it's needed.

File: src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        """
+        Pilot log getter method, carrying the unique pilot identity and a VO name.
+
+        :param str pilotStamp: pilot stamp.
+        :param str vo: VO name of a user/pilot which generated the logs.
+        :return: S_OK or S_ERROR
+        :rtype: dict
+        """
+
+        opsHelper = Operations(vo=vo, setup=self.setup)
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py]:
```suggestion
        opsHelper = Operations(vo=vo)
```

File: src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        res = self._downloadLogs(  # pylint: disable=unexpected-keyword-arg
+            lfn, filepath, proxyUserName=proxyUser, proxyUserGroup=proxyGroup
+        )
+        sLog.debug("getFile result:", res)
+        if not res["OK"]:
+            sLog.error(f"Failed to contact storage")
+            return res
+        if lfn in res["Value"]["Failed"]:
+            sLog.error("Failed to retrieve a log file:", res["Value"]["Failed"])
+            return res
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py]: Don't you want to return an `S_ERROR` here?

File: src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        :param str pilotStamp: pilot stamp.
+        :param str vo: VO name of a user/pilot which generated the logs.
+        :return: S_OK or S_ERROR
+        :rtype: dict
+        """
+
+        opsHelper = Operations(vo=vo, setup=self.setup)
+        uploadPath = opsHelper.getValue("Pilot/UploadPath", "")
+        lfn = os.path.join(uploadPath, pilotStamp + ".log")
+        sLog.info("LFN to download: ", lfn)
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py:42]: Is this really an LFN? Looks like just a path to a file to me.

File: src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            filename = os.path.join(filepath, pilotStamp + ".log")
+            with open(filename) as f:
+                stdout = f.read()
+        except FileNotFoundError as err:
+            sLog.error(f"Error opening a log file:{filename}", err)
+            return S_ERROR(repr(err))
+
+        resultDict = {}
+        resultDict["StdOut"] = stdout
+        return S_OK(resultDict)
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py:89]: Why not simply
 ```suggestion
        return S_OK(stdout)
```
?

File: src/DIRAC/WorkloadManagementSystem/Service/PilotManagerHandler.py [hunk 8]
```diff
@@ -152,6 +158,54 @@ def export_getPilotOutput(self, pilotReference):
...
+        :param str pilotReference:
+        :return: S_OK Dirac object
+        :rtype: dict
+        """
+
+        # get pilot stamp by pilot reference:
+        result = self.pilotAgentsDB.getPilotInfo(pilotReference)
+        if not result["OK"]:
+            self.log.error("Failed to get info for pilot", result["Message"])
+            return S_ERROR("Failed to get info for pilot")
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Service/PilotManagerHandler.py]:
```suggestion
            return res
```

File: src/DIRAC/WorkloadManagementSystem/Service/PilotManagerHandler.py [hunk 9]
```diff
@@ -152,6 +158,54 @@ def export_getPilotOutput(self, pilotReference):
...
+
+            componentClass = result["Value"]
+            self.loggingPlugin = componentClass()
+            self.log.info("Loaded: PilotLoggingPlugin class", self.configValue)
+
+        res = self.loggingPlugin.getRemotePilotLogs(pilotStamp, vo)
+
+        if res["OK"]:
+            res["Value"]["OwnerDN"] = owner
+            res["Value"]["OwnerGroup"] = group
```
Comment (fstagni) [inline src/DIRAC/WorkloadManagementSystem/Service/PilotManagerHandler.py]: What is the use of these values?

Comment (martynia) [reply src/DIRAC/WorkloadManagementSystem/Service/PilotManagerHandler.py]: A good question. The classic log retriever does the same here:
https://github.com/DIRACGrid/DIRAC/blob/bef4e37ba9e3f936fa6787476e2fd0a3a3ac17d6/src/DIRAC/WorkloadManagementSystem/Service/PilotManagerHandler.py#L99

So I did the same, although, I can't see the use of it

File: src/DIRAC/WorkloadManagementSystem/Service/PilotManagerHandler.py (hunk 8 above)
Comment (martynia) [reply src/DIRAC/WorkloadManagementSystem/Service/PilotManagerHandler.py]: Sure.

File: src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py (hunk 7 above)
Comment (martynia) [reply src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py:89]: This is required by a client:

https://github.com/DIRACGrid/DIRAC/blob/bef4e37ba9e3f936fa6787476e2fd0a3a3ac17d6/src/DIRAC/Interfaces/API/DiracAdmin.py#L469

File: src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py (hunk 5 above)
Comment (martynia) [reply src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py]: Yes

File: src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py (hunk 6 above)
Comment (martynia) [reply src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py:42]: It is. It is passed to `getFile` later

File: src/DIRAC/WorkloadManagementSystem/Service/PilotManagerHandler.py (hunk 9 above)
Comment (fstagni) [reply src/DIRAC/WorkloadManagementSystem/Service/PilotManagerHandler.py]: Considering that in https://github.com/DIRACGrid/DIRAC/pull/7157 I have removed OwnerDN, do not use it here.

File: src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py (hunk 6 above)
Comment (fstagni) [reply src/DIRAC/WorkloadManagementSystem/Client/PilotLoggingPlugins/FileCacheDownloadPlugin.py:42]: I am afraid we have been using this pattern in other places, but `os.path` is not the most precise way to manipulate/create LFNs, as this is OS-dependent. We might argue that we only run on Linux so it does not really matter, but more precise would be to at least to use `posixpath` instead (@chaen and @atsareg might comment)

File: src/DIRAC/WorkloadManagementSystem/Service/PilotManagerHandler.py (hunk 9 above)
Comment (martynia) [reply src/DIRAC/WorkloadManagementSystem/Service/PilotManagerHandler.py]: OK

**Review by fstagni — APPROVED** [review]

**Pull request merged**